README.md
[![Build Status](https://ci-cd.springdoc.org:8443/buildStatus/icon?job=springdoc-openapi-gradle-IC)](https://ci-cd.springdoc.org:8443/view/springdoc-openapi/job/springdoc-openapi-gradle-IC/)

# Introducing springdoc-openapi-gradle-plugin

Gradle plugin for springdoc-openapi.

This plugin allows you to generate an OpenAPI 3 specification for a Spring Boot
application from a Gradle build.
Compatibility Notes
-------------------

The plugin is built on Gradle version 7.0.

Dependencies
------------
This plugin has a runtime dependency on the following plugins:

1. Spring Boot Gradle plugin - `org.springframework.boot`
2. Gradle process plugin - `com.github.psxpaul.execfork`

How To Use
----------

Gradle Groovy DSL

```groovy
plugins {
      id "org.springframework.boot" version "2.7.0"
      id "org.springdoc.openapi-gradle-plugin" version "1.7.0"
}
```

Gradle Kotlin DSL

```groovy
plugins {
    id("org.springframework.boot") version "2.7.0"
    id("org.springdoc.openapi-gradle-plugin") version "1.7.0"
}
```

Note: For latest versions of the plugins please check
the [Gradle Plugins portal](https://plugins.gradle.org/).

How the plugin works?
------------

When you add this plugin and its runtime dependency plugins to your build file, the plugin
creates the following tasks:

1. forkedSpringBootRun

2. generateOpenApiDocs

Running the task `generateOpenApiDocs` writes the OpenAPI spec into a `openapi.json` file
in your project's build dir.

```bash
gradle clean generateOpenApiDocs
``` 

When you run the gradle task **generateOpenApiDocs**, it starts your spring boot
application in the background using **forkedSpringBootRun** task.
Once your application is up and running **generateOpenApiDocs** makes a rest call to your
applications doc url to download and store the open api docs file as json.


Customization
-------------

The following customizations can be done on task generateOpenApiDocs using extension
openApi as follows

```kotlin
openApi {
    apiDocsUrl.set("https://localhost:9000/api/docs")
    outputDir.set(file("$buildDir/docs"))
    outputFileName.set("swagger.json")
    waitTimeInSeconds.set(10)
    groupedApiMappings.set(["https://localhost:8080/v3/api-docs/groupA" to "swagger-groupA.json",
                            "https://localhost:8080/v3/api-docs/groupB" to "swagger-groupB.json"])
    customBootRun {
        args.set(["--spring.profiles.active=special"]) 
    }
}
```

| Parameter            | Description                                                                                                                         | Required | Default                              |
|----------------------|-------------------------------------------------------------------------------------------------------------------------------------|----------|--------------------------------------|
| `apiDocsUrl`         | The URL from where the OpenAPI doc can be downloaded. If the url ends with `.yaml`, output will YAML format.                                                                                 | No       | http://localhost:8080/v3/api-docs    |
| `outputDir`          | The output directory for the generated OpenAPI file                                                                                 | No       | $buildDir - Your project's build dir |
| `outputFileName`     | Specifies the output file name.                            | No       | openapi.json                         |
| `waitTimeInSeconds`  | Time to wait in seconds for your Spring Boot application to start, before we make calls to `apiDocsUrl` to download the OpenAPI doc | No       | 30 seconds                           |
| `groupedApiMappings` | A map of URLs (from where the OpenAPI docs can be downloaded) to output file names                                                  | No       | []                                   |
| `customBootRun`      | Any bootRun property that you would normal need to start your spring boot application.                                              | No       | (N/A)                                |

### `customBootRun` properties examples

`customBootRun` allows you to send in the properties that might be necessary to allow for
the forked spring boot application that gets started
to be able to start (profiles, other custom properties, etc.)
`customBootRun` allows you can specify bootRun style parameter, such
as `args`, `jvmArgs`, `systemProperties` and `workingDir`.
If you don't specify `customBootRun` parameter, this plugin uses the parameter specified
to `bootRun` in Spring Boot Gradle Plugin.

#### Passing static args

This allows for you to be able to just send the static properties when executing Spring
application in `generateOpenApiDocs`.

```
openApi {
    customBootRun {
        args = ["--spring.profiles.active=special"] 
    }
}
```

#### Passing straight from gradle

This allows for you to be able to just send in whatever you need when you generate docs.

`./gradlew clean generateOpenApiDocs -Dspring.profiles.active=special`

and as long as the config looks as follows that value will be passed into the forked
spring boot application.

```
openApi {
    customBootRun {
         systemProperties = System.properties
    }
}
```

### Grouped API Mappings Notes

The `groupedApiMappings` customization allows you to specify multiple URLs/file names for
use within this plugin. This configures the plugin to ignore the `apiDocsUrl`
and `outputFileName` parameters and only use those found in `groupedApiMappings`. The
plugin will then attempt to download each OpenAPI doc in turn as it would for a single
OpenAPI doc.

# Building the plugin

1. Clone the repo `git@github.com:springdoc/springdoc-openapi-gradle-plugin.git`
2. Build and publish the plugin into your local maven repository by running the following
    ```
    ./gradlew clean pTML
   ```

# Testing the plugin

1. Create a new spring boot application or use an existing spring boot app and follow
   the `How To Use` section above to configure this plugin.
2. Update the version for the plugin to match the current version found
   in `build.gradle.kts`

    ```
    id("org.springdoc.openapi-gradle-plugin") version "1.7.0"
    ```

3. Add the following to the spring boot apps `settings.gradle`

    ```
    pluginManagement {
        repositories {
            mavenLocal()
            gradlePluginPortal()
        }
    }
    ```

# **Thank you for the support**

* Thanks a lot [JetBrains](https://www.jetbrains.com/?from=springdoc-openapi) for
  supporting springdoc-openapi project.

![JetBrains logo](https://springdoc.org/img/jetbrains.svg)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [UnReleased] -

## Added


src/test/kotlin/org/springdoc/openapi/gradle/plugin/OpenApiGradlePluginTest.kt
package org.springdoc.openapi.gradle.plugin

import com.beust.klaxon.JsonObject
import com.beust.klaxon.Parser
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.gradle.internal.impldep.org.apache.commons.lang.RandomStringUtils
import org.gradle.testkit.runner.BuildResult
import org.gradle.testkit.runner.GradleRunner
import org.gradle.testkit.runner.TaskOutcome
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.io.FileReader
import java.nio.file.Files

class OpenApiGradlePluginTest {

	private val projectTestDir = Files.createTempDirectory("acceptance-project").toFile()
	private val buildFile = File(projectTestDir, "build.gradle")
	private val projectBuildDir = File(projectTestDir, "build")

	private val pathsField = "paths"
	private val openapiField = "openapi"

	private val baseBuildGradle = """plugins {
            id 'java'
            id 'org.springframework.boot' version '2.7.6'
            id 'io.spring.dependency-management' version '1.0.15.RELEASE'
            id 'org.springdoc.openapi-gradle-plugin'
        }
        
        group = 'com.example'
        version = '0.0.1-SNAPSHOT'
        sourceCompatibility = '8'
        
        repositories {
            mavenCentral()
        }
        
        dependencies {
            implementation 'org.springframework.boot:spring-boot-starter-web'
            implementation 'org.springdoc:springdoc-openapi-webmvc-core:1.6.13'
        }
    """.trimIndent()

	companion object {
		val logger: Logger = LoggerFactory.getLogger(OpenApiGradlePluginTest::class.java)
	}

	@BeforeEach
	fun createTemporaryAcceptanceProjectFromTemplate() {
		File(javaClass.classLoader.getResource("acceptance-project")!!.path).copyRecursively(
			projectTestDir
		)
	}

	@Test
	fun `default build no options`() {
		buildFile.writeText(baseBuildGradle)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1)
	}

	@Test
	fun `different output dir`() {
		val specialOutputDir = File(projectTestDir, "specialDir")
		specialOutputDir.mkdirs()

		buildFile.writeText(
			"""$baseBuildGradle
            openApi{
                outputDir = file("${specialOutputDir.toURI().path}")
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1, buildDir = specialOutputDir)
	}

	@Test
	fun `different output file name`() {
		val specialOutputJsonFileName = RandomStringUtils.randomAlphanumeric(15)

		buildFile.writeText(
			"""$baseBuildGradle
            openApi{
                outputFileName = "$specialOutputJsonFileName"
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1, specialOutputJsonFileName)
	}

	@Test
	fun `using properties`() {
		buildFile.writeText(
			"""$baseBuildGradle
            bootRun {
                args = ["--spring.profiles.active=multiple-endpoints", "--some.second.property=someValue"]
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(3)
	}

	@Test
	fun `using forked properties via System properties`() {
		buildFile.writeText(
			"""$baseBuildGradle
            bootRun {
                systemProperties = System.properties
            }
        """.trimMargin()
		)

		assertEquals(
			TaskOutcome.SUCCESS,
			openApiDocsTask(runTheBuild("-Dspring.profiles.active=multiple-endpoints")).outcome
		)
		assertOpenApiJsonFile(2)
	}

	@Test
	fun `using forked properties via System properties with customBootRun`() {
		buildFile.writeText(
			"""$baseBuildGradle
            openApi {
                customBootRun {
                    systemProperties = System.properties
                }
            }
        """.trimMargin()
		)

		assertEquals(
			TaskOutcome.SUCCESS,
			openApiDocsTask(runTheBuild("-Dspring.profiles.active=multiple-endpoints")).outcome
		)
		assertOpenApiJsonFile(2)
	}

	@Test
	fun `configurable wait time`() {
		buildFile.writeText(
			"""$baseBuildGradle
           bootRun {
                args = ["--spring.profiles.active=slower"]
            }
             openApi{
                waitTimeInSeconds = 60
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1)
	}

	@Test
	fun `using different api url`() {
		buildFile.writeText(
			"""$baseBuildGradle
           bootRun {
                args = ["--spring.profiles.active=different-url"]
            }
            openApi{
                apiDocsUrl = "http://localhost:8080/secret-api-docs"
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1)
	}

	@Test
	fun `using different api url via customBootRun`() {
		buildFile.writeText(
			"""$baseBuildGradle
            openApi{
                apiDocsUrl = "http://localhost:8080/secret-api-docs"
                customBootRun {
                    args = ["--spring.profiles.active=different-url"]
                }
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1)
	}

	@Test
	fun `yaml generation`() {
		val outputYamlFileName = "openapi.yaml"

		buildFile.writeText(
			"""$baseBuildGradle
                
            openApi{
                apiDocsUrl = "http://localhost:8080/v3/api-docs.yaml"
                outputFileName = "$outputYamlFileName"
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiYamlFile(1, outputYamlFileName)
	}

	@Test
	fun `using multiple grouped apis`() {
		val outputJsonFileNameGroupA = "openapi-groupA.json"
		val outputJsonFileNameGroupB = "openapi-groupB.json"

		buildFile.writeText(
			"""$baseBuildGradle
           bootRun {
                args = ["--spring.profiles.active=multiple-grouped-apis"]
            }
            openApi{
                groupedApiMappings = ["http://localhost:8080/v3/api-docs/groupA": "$outputJsonFileNameGroupA",
                                      "http://localhost:8080/v3/api-docs/groupB": "$outputJsonFileNameGroupB"]
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiJsonFile(1, outputJsonFileNameGroupA)
		assertOpenApiJsonFile(2, outputJsonFileNameGroupB)
	}

	@Test
	fun `using multiple grouped apis with yaml`() {
		val outputYamlFileNameGroupA = "openapi-groupA.yaml"
		val outputYamlFileNameGroupB = "openapi-groupB.yaml"

		buildFile.writeText(
			"""$baseBuildGradle
           bootRun {
                args = ["--spring.profiles.active=multiple-grouped-apis"]
            }
            openApi{
                groupedApiMappings = ["http://localhost:8080/v3/api-docs.yaml/groupA": "$outputYamlFileNameGroupA",
                                      "http://localhost:8080/v3/api-docs.yaml/groupB": "$outputYamlFileNameGroupB"]
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertOpenApiYamlFile(1, outputYamlFileNameGroupA)
		assertOpenApiYamlFile(2, outputYamlFileNameGroupB)
	}

	@Test
	fun `using multiple grouped apis should ignore single api properties`() {
		val outputJsonFileNameSingleGroupA = "openapi-single-groupA.json"
		val outputJsonFileNameGroupA = "openapi-groupA.json"
		val outputJsonFileNameGroupB = "openapi-groupB.json"

		buildFile.writeText(
			"""$baseBuildGradle
           bootRun {
                args = ["--spring.profiles.active=multiple-grouped-apis"]
            }
            openApi{
                apiDocsUrl = "http://localhost:8080/v3/api-docs/groupA"
                outputFileName = "$outputJsonFileNameSingleGroupA"
                groupedApiMappings = ["http://localhost:8080/v3/api-docs/groupA": "$outputJsonFileNameGroupA",
                                      "http://localhost:8080/v3/api-docs/groupB": "$outputJsonFileNameGroupB"]
            }
        """.trimMargin()
		)

		assertEquals(TaskOutcome.SUCCESS, openApiDocsTask(runTheBuild()).outcome)
		assertFalse(File(projectBuildDir, outputJsonFileNameSingleGroupA).exists())
		assertOpenApiJsonFile(1, outputJsonFileNameGroupA)
		assertOpenApiJsonFile(2, outputJsonFileNameGroupB)
	}

	@Test
	fun `using invalid doc url`() {
		buildFile.writeText(
			"""$baseBuildGradle
            openApi{
                apiDocsUrl = "http://localhost:8080/hello/world"
            }
        """.trimMargin()
		)

		try {
			openApiDocsTask(runTheBuild())
		} catch (e: RuntimeException) {
			logger.error(e.message)
			assertNotNull(e.message?.lines()?.find {
				it.contains(
					"Failed to parse the API docs response string. " +
							"Please ensure that the response is in the correct format."
				)
			})
		}
	}

	private fun runTheBuild(vararg additionalArguments: String = emptyArray()) =
		GradleRunner.create()
			.withProjectDir(projectTestDir)
			.withArguments("clean", "generateOpenApiDocs", *additionalArguments)
			.withPluginClasspath()
			.build()

	private fun assertOpenApiJsonFile(
		expectedPathCount: Int,
		outputJsonFileName: String = DEFAULT_OPEN_API_FILE_NAME,
		buildDir: File = projectBuildDir
	) {
		val openApiJson = getOpenApiJsonAtLocation(File(buildDir, outputJsonFileName))
		assertEquals("3.0.1", openApiJson.string(openapiField))
		assertEquals(expectedPathCount, openApiJson.obj(pathsField)!!.size)
	}

	private fun getOpenApiJsonAtLocation(path: File) =
		Parser.default().parse(FileReader(path)) as JsonObject

	private fun assertOpenApiYamlFile(
		expectedPathCount: Int,
		outputJsonFileName: String = DEFAULT_OPEN_API_FILE_NAME,
		buildDir: File = projectBuildDir
	) {
		val mapper = ObjectMapper(YAMLFactory())
		mapper.registerModule(KotlinModule.Builder().build())
		val node = mapper.readTree(File(buildDir, outputJsonFileName))
		assertEquals("3.0.1", node.get(openapiField).asText())
		assertEquals(expectedPathCount, node.get(pathsField)!!.size())
	}

	private fun openApiDocsTask(result: BuildResult) =
		result.tasks.find { it.path.contains("generateOpenApiDocs") }!!
}


src/test/resources/acceptance-project/src/main/java/com/example/demo/DemoApplication.java
package com.example.demo;

import java.time.Duration;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class DemoApplication implements ApplicationRunner {

	@Value("${slower:false}")
	boolean slower;

	public static void main(String[] args) {
		SpringApplication.run(DemoApplication.class, args);
	}

	@Override
	public void run(ApplicationArguments arg0) throws Exception {
		System.out.println("Hello World from Application Runner");
		if (slower) {
			Duration waitTime = Duration.ofSeconds(40);
			System.out.println("Waiting for " + waitTime + " before starting");
			Thread.sleep(waitTime.toMillis());
		}
	}

}


src/test/resources/acceptance-project/src/main/java/com/example/demo/config/GroupedConfiguration.java
package com.example.demo;

import org.springdoc.core.GroupedOpenApi;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

@Profile("multiple-grouped-apis")
@Configuration
public class GroupedConfiguration {

	@Bean
	public GroupedOpenApi groupA() {
		return GroupedOpenApi.builder()
				.group("groupA")
				.pathsToMatch("/groupA/**")
				.build();
	}

	@Bean
	public GroupedOpenApi groupB() {
		return GroupedOpenApi.builder()
				.group("groupB")
				.pathsToMatch("/groupB/**")
				.build();
	}
}


src/main/kotlin/org/springdoc/openapi/gradle/plugin/AnnotatedFork.kt
package org.springdoc.openapi.gradle.plugin

import com.github.jengelman.gradle.plugins.processes.tasks.Fork
import groovy.lang.MetaClass
import org.codehaus.groovy.runtime.InvokerHelper
import org.gradle.api.tasks.Internal
import org.gradle.process.CommandLineArgumentProvider

/**
 * Extends the Fork task from the gradle-processes plugin to be compatible with Gradle 7.
 * The functionality of the class remains exactly the same; it was just missing an annotation
 * on the [getArgumentProviders] method and the annotation is mandatory in Gradle 7.
 *
 * Since the original class is written in Groovy, we also need to implement some internal
 * Groovy methods that get added to Groovy classes via bytecode manipulation.
 */
open class AnnotatedFork : Fork() {
    private var metaClass: MetaClass? = null

    override fun invokeMethod(method: String?, args: Any?): Any? {
        return InvokerHelper.invokeMethod(this, method, args)
    }

    override fun getProperty(property: String?): Any? {
        return InvokerHelper.getProperty(this, property)
    }

    override fun getMetaClass(): MetaClass? {
        if (metaClass == null) {
            metaClass = InvokerHelper.getMetaClass(javaClass)
        }
        return metaClass
    }

    override fun setMetaClass(newMetaClass: MetaClass?) {
        metaClass = newMetaClass
    }

    @Internal
    override fun getArgumentProviders(): MutableList<CommandLineArgumentProvider>? {
        return super.getArgumentProviders()
    }
}


src/main/kotlin/org/springdoc/openapi/gradle/plugin/OpenApiExtension.kt
package org.springdoc.openapi.gradle.plugin

import org.gradle.api.Action
import org.gradle.api.file.ConfigurableFileCollection
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.ProjectLayout
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.ListProperty
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import javax.inject.Inject

open class OpenApiExtension @Inject constructor(
	objects: ObjectFactory,
	layout: ProjectLayout
) {
	val apiDocsUrl: Property<String> = objects.property(String::class.java)
	val outputFileName: Property<String> = objects.property(String::class.java)
	val outputDir: DirectoryProperty = objects.directoryProperty()
	val waitTimeInSeconds: Property<Int> = objects.property(Int::class.java)
	val groupedApiMappings: MapProperty<String, String> =
		objects.mapProperty(String::class.java, String::class.java)
	val customBootRun: CustomBootRunAction =
		objects.newInstance(CustomBootRunAction::class.java)

	init {
		apiDocsUrl.convention(DEFAULT_API_DOCS_URL)
		outputFileName.convention(DEFAULT_OPEN_API_FILE_NAME)
		outputDir.convention(layout.buildDirectory)
		waitTimeInSeconds.convention(DEFAULT_WAIT_TIME_IN_SECONDS)
		groupedApiMappings.convention(emptyMap())
	}

	fun customBootRun(action: Action<CustomBootRunAction>) {
		action.execute(customBootRun)
	}
}

open class CustomBootRunAction @Inject constructor(
	objects: ObjectFactory,
) {
	val systemProperties: MapProperty<String, Any> =
		objects.mapProperty(String::class.java, Any::class.java)
	val workingDir: DirectoryProperty = objects.directoryProperty()
	val mainClass: Property<String> = objects.property(String::class.java)
	val args: ListProperty<String> = objects.listProperty(String::class.java)
	val classpath: ConfigurableFileCollection = objects.fileCollection()
	val jvmArgs: ListProperty<String> = objects.listProperty(String::class.java)
	val environment: MapProperty<String, Any> =
		objects.mapProperty(String::class.java, Any::class.java)
}


src/main/kotlin/org/springdoc/openapi/gradle/plugin/OpenApiGradlePlugin.kt
package org.springdoc.openapi.gradle.plugin

import com.github.psxpaul.task.JavaExecFork
import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.Task
import org.gradle.api.UnknownDomainObjectException
import org.gradle.api.tasks.TaskProvider
import org.gradle.internal.jvm.Jvm
import org.springframework.boot.gradle.tasks.run.BootRun

open class OpenApiGradlePlugin : Plugin<Project> {

	override fun apply(project: Project) {
		with(project) {
			// Run time dependency on the following plugins
			plugins.apply(SPRING_BOOT_PLUGIN)
			plugins.apply(EXEC_FORK_PLUGIN)

			extensions.create(EXTENSION_NAME, OpenApiExtension::class.java)
			tasks.register(FORKED_SPRING_BOOT_RUN_TASK_NAME, JavaExecFork::class.java)
			tasks.register(OPEN_API_TASK_NAME, OpenApiGeneratorTask::class.java)

			generate(this)
		}
	}

	private fun generate(project: Project) = project.run {
		springBoot3CompatibilityCheck()

		// The task, used to run the Spring Boot application (`bootRun`)
		val bootRunTask = tasks.named(SPRING_BOOT_RUN_TASK_NAME)
		// The task, used to resolve the application's main class (`bootRunMainClassName`)
		val bootRunMainClassNameTask =
			try {
				val task = tasks.named(SPRING_BOOT_RUN_MAIN_CLASS_NAME_TASK_NAME)
				logger.debug(
					"Detected Spring Boot task {}",
					SPRING_BOOT_RUN_MAIN_CLASS_NAME_TASK_NAME
				)
				task
			} catch (e: UnknownDomainObjectException) {
				val task = tasks.named(SPRING_BOOT_3_RUN_MAIN_CLASS_NAME_TASK_NAME)
				logger.debug(
					"Detected Spring Boot task {}",
					SPRING_BOOT_3_RUN_MAIN_CLASS_NAME_TASK_NAME
				)
				task
			}

		val extension = extensions.findByName(EXTENSION_NAME) as OpenApiExtension
		val customBootRun = extension.customBootRun
		// Create a forked version spring boot run task
		val forkedSpringBoot = tasks.named(
			FORKED_SPRING_BOOT_RUN_TASK_NAME,
			JavaExecFork::class.java
		) { fork ->
			fork.dependsOn(tasks.named(bootRunMainClassNameTask.name))
			fork.onlyIf { needToFork(bootRunTask, customBootRun, fork) }
		}

		// This is my task. Before I can run it, I have to run the dependent tasks
		val openApiTask =
			tasks.named(OPEN_API_TASK_NAME, OpenApiGeneratorTask::class.java) {
				it.dependsOn(forkedSpringBoot)
			}

		// The forked task need to be terminated as soon as my task is finished
		forkedSpringBoot.get().stopAfter = openApiTask as TaskProvider<Task>
	}

	private fun Project.springBoot3CompatibilityCheck() {
		val tasksNames = tasks.names
		val boot2TaskName = "bootRunMainClassName"
		val boot3TaskName = "resolveMainClassName"
		if (!tasksNames.contains(boot2TaskName) && tasksNames.contains(boot3TaskName)) {
		    tasks.register(boot2TaskName) { it.dependsOn(tasks.named(boot3TaskName)) }
		}
	}

	private fun needToFork(
		bootRunTask: TaskProvider<Task>,
		customBootRun: CustomBootRunAction,
		fork: JavaExecFork
	): Boolean {
		val bootRun = bootRunTask.get() as BootRun

		val baseSystemProperties =
			customBootRun.systemProperties.orNull?.takeIf { it.isNotEmpty() }
				?: bootRun.systemProperties
		with(fork) {
			// copy all system properties, excluding those starting with `java.class.path`
			systemProperties = baseSystemProperties.filter {
				!it.key.startsWith(CLASS_PATH_PROPERTY_NAME)
			}

			// use original bootRun parameter if the list-type customBootRun properties are empty
			workingDir = customBootRun.workingDir.asFile.orNull
				?: fork.workingDir
			args = customBootRun.args.orNull?.takeIf { it.isNotEmpty() }?.toMutableList()
				?: bootRun.args?.toMutableList() ?: mutableListOf()
			classpath = customBootRun.classpath.takeIf { !it.isEmpty }
				?: bootRun.classpath
			main = customBootRun.mainClass.orNull
				?: bootRun.mainClass.get()
			jvmArgs = customBootRun.jvmArgs.orNull?.takeIf { it.isNotEmpty() }
				?: bootRun.jvmArgs
			environment = customBootRun.environment.orNull?.takeIf { it.isNotEmpty() }
				?: bootRun.environment
			if (Jvm.current().toString().startsWith("1.8")) {
				killDescendants = false
			}
		}
		return true
	}
}


src/main/kotlin/org/springdoc/openapi/gradle/plugin/OpenApiGeneratorTask.kt
package org.springdoc.openapi.gradle.plugin

import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonSyntaxException
import org.awaitility.Durations
import org.awaitility.core.ConditionTimeoutException
import org.awaitility.kotlin.atMost
import org.awaitility.kotlin.await
import org.awaitility.kotlin.ignoreException
import org.awaitility.kotlin.until
import org.awaitility.kotlin.withPollInterval
import org.gradle.api.DefaultTask
import org.gradle.api.GradleException
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.provider.MapProperty
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.TaskAction
import java.net.ConnectException
import java.net.HttpURLConnection
import java.net.URL
import java.time.Duration
import java.time.temporal.ChronoUnit.SECONDS
import java.util.*

private const val MAX_HTTP_STATUS_CODE = 299

open class OpenApiGeneratorTask : DefaultTask() {
	@get:Input
	val apiDocsUrl: Property<String> = project.objects.property(String::class.java)

	@get:Input
	val outputFileName: Property<String> = project.objects.property(String::class.java)

	@get:Input
	val groupedApiMappings: MapProperty<String, String> =
		project.objects.mapProperty(String::class.java, String::class.java)

	@get:OutputDirectory
	val outputDir: DirectoryProperty = project.objects.directoryProperty()
	private val waitTimeInSeconds: Property<Int> =
		project.objects.property(Int::class.java)

	init {
		description = OPEN_API_TASK_DESCRIPTION
		group = GROUP_NAME
		// load my extensions
		val extension: OpenApiExtension =
			project.extensions.getByName(EXTENSION_NAME) as OpenApiExtension

		// set a default value if not provided
		val defaultOutputDir = project.objects.directoryProperty()
		defaultOutputDir.set(project.buildDir)

		apiDocsUrl.convention(extension.apiDocsUrl.getOrElse(DEFAULT_API_DOCS_URL))
		outputFileName.convention(
			extension.outputFileName.getOrElse(
				DEFAULT_OPEN_API_FILE_NAME
			)
		)
		groupedApiMappings.convention(extension.groupedApiMappings.getOrElse(emptyMap()))
		outputDir.convention(extension.outputDir.getOrElse(defaultOutputDir.get()))
		waitTimeInSeconds.convention(
			extension.waitTimeInSeconds.getOrElse(
				DEFAULT_WAIT_TIME_IN_SECONDS
			)
		)
	}

	@TaskAction
	fun execute() {
		if (groupedApiMappings.isPresent && groupedApiMappings.get().isNotEmpty()) {
			groupedApiMappings.get().forEach(this::generateApiDocs)
		} else {
			generateApiDocs(apiDocsUrl.get(), outputFileName.get())
		}
	}

	private fun generateApiDocs(url: String, fileName: String) {
		try {
			val isYaml = url.toLowerCase().matches(Regex(".+[./]yaml(/.+)*"))
			await ignoreException ConnectException::class withPollInterval Durations.ONE_SECOND atMost Duration.of(
				waitTimeInSeconds.get().toLong(),
				SECONDS
			) until {
				val connection: HttpURLConnection =
					URL(url).openConnection() as HttpURLConnection
				connection.requestMethod = "GET"
				connection.connect()
				val statusCode = connection.responseCode
				logger.trace("apiDocsUrl = {} status code = {}", url, statusCode)
				statusCode < MAX_HTTP_STATUS_CODE
			}
			logger.info("Generating OpenApi Docs..")
			val connection: HttpURLConnection =
				URL(url).openConnection() as HttpURLConnection
			connection.requestMethod = "GET"
			connection.connect()

			val response = String(connection.inputStream.readBytes(), Charsets.UTF_8)

			val apiDocs = if (isYaml) response else prettifyJson(response)

			val outputFile = outputDir.file(fileName).get().asFile
			outputFile.writeText(apiDocs)
		} catch (e: ConditionTimeoutException) {
			this.logger.error(
				"Unable to connect to $url waited for ${waitTimeInSeconds.get()} seconds",
				e
			)
			throw GradleException("Unable to connect to $url waited for ${waitTimeInSeconds.get()} seconds")
		}
	}

	private fun prettifyJson(response: String): String {
		val gson = GsonBuilder().setPrettyPrinting().create()
		try {
			val googleJsonObject = gson.fromJson(response, JsonObject::class.java)
			return gson.toJson(googleJsonObject)
		} catch (e: RuntimeException) {
			throw JsonSyntaxException(
				"Failed to parse the API docs response string. " +
						"Please ensure that the response is in the correct format. response=$response",
				e
			)
		}
	}
}


src/main/kotlin/org/springdoc/openapi/gradle/plugin/Constants.kt
package org.springdoc.openapi.gradle.plugin

const val EXTENSION_NAME = "openApi"
const val GROUP_NAME = "OpenApi"
const val OPEN_API_TASK_NAME = "generateOpenApiDocs"
const val OPEN_API_TASK_DESCRIPTION = "Generates the spring doc openapi file"
const val SPRING_BOOT_RUN_TASK_NAME = "bootRun"
const val SPRING_BOOT_RUN_MAIN_CLASS_NAME_TASK_NAME = "bootRunMainClassName"
const val SPRING_BOOT_3_RUN_MAIN_CLASS_NAME_TASK_NAME = "bootRun"
const val FORKED_SPRING_BOOT_RUN_TASK_NAME = "forkedSpringBootRun"
const val INSPECT_IC_CLASSES_TASK_NAME = "inspectClassesForKotlinIC"

const val DEFAULT_API_DOCS_URL = "http://localhost:8080/v3/api-docs"
const val DEFAULT_OPEN_API_FILE_NAME = "openapi.json"
const val DEFAULT_WAIT_TIME_IN_SECONDS = 30

const val SPRING_BOOT_PLUGIN = "org.springframework.boot"
const val EXEC_FORK_PLUGIN = "com.github.psxpaul.execfork"

const val CLASS_PATH_PROPERTY_NAME = "java.class.path"
